private/bufpkg/bufconnect/interceptors.go

```diff
@@ -28,17 +30,33 @@ const (
 // NewSetCLIVersionInterceptor returns a new Connect Interceptor that sets the Buf CLI version into all request headers
 func NewSetCLIVersionInterceptor(version string) connect.UnaryInterceptorFunc {
 	interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-		return func(
-			ctx context.Context,
-			req connect.AnyRequest,
-		) (connect.AnyResponse, error) {
+		return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
 			req.Header().Set(CliVersionHeaderName, version)
 			return next(ctx, req)
 		}
 	}
 	return interceptor
 }
 
+// NewCLIWarningInterceptor returns a new Connect Interceptor that logs CLI warnings returned by sever responses.
+func NewCLIWarningInterceptor(container applog.Container) connect.UnaryInterceptorFunc {
+	interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
+		return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+			resp, err := next(ctx, req)
+			if resp != nil && resp.Header() != nil {
+				encoded := resp.Header().Get(CLIWarningHeaderName)
+				if encoded != "" {
+					if warning, err := base64.StdEncoding.DecodeString(encoded); err == nil && len(warning) > 0 {
```

Comment:
To be fully safe, headers must be ASCII. That makes it annoying to include problematic user-supplied data in the error message (e.g., "username 💩 is not a valid Kafka topic").

Reply:
Covers any future case of allowing non-ASCII strings. If we don't do it, we'll probably regret it in the future for older CLI versions, and there is no downside outside of not seeing the raw text in the header. Since this is intended for the CLI, it seems like a reasonable tradeoff.

Reply:
OK got it

Reply:
Should we use [connect.DecodeBinaryHeader](https://pkg.go.dev/github.com/bufbuild/connect-go#DecodeBinaryHeader) instead?

Relatedly:

> Following usual HTTP semantics, multiple base64-encoded values may be joined with a comma. When receiving such comma-separated values, split them with [strings.Split](https://pkg.go.dev/strings#Split) before calling DecodeBinaryHeader.

Reply:
Fair suggestion, updated in https://github.com/bufbuild/buf/pull/2171/commits/ea0d65ad6ad7cfee95719cb225cdf0407da65966